### ISSUE TITLE
Phase 33.4 — CurriculumScheduler: Intelligent task sequencing for lifelong learning

### ISSUE THREAD

**Issue by web3guru888** [web3guru888/asi-build#694]
## Overview

The `CurriculumScheduler` manages the ordering and timing of task presentation in continual learning, optimizing both forward transfer (learning new tasks faster by leveraging previous knowledge) and backward transfer (improving old tasks through new learning). It implements curriculum learning strategies including difficulty-based ordering, spaced repetition, interleaving, and anti-curriculum detection to identify harmful task sequences.

## Key Responsibilities

- **Task difficulty estimation** — Score tasks by complexity using model confidence, loss landscape curvature, or gradient norms
- **Optimal ordering** — Sequence tasks to maximize positive forward transfer using task similarity graphs
- **Spaced repetition** — Schedule review of previous tasks at exponentially increasing intervals (Leitner system adapted for neural CL)
- **Interleaving strategies** — Mix tasks within training epochs to improve generalization
- **Anti-curriculum detection** — Identify task orderings that cause excessive negative backward transfer
- **Transfer matrix tracking** — Maintain running estimates of pairwise forward/backward transfer between all tasks
- **Dynamic reordering** — Adjust remaining task schedule based on observed transfer during training

### Inputs
- `task_pool: List[TaskSpecification]` — Available tasks with metadata
- `ordering_strategy: Literal["difficulty_asc", "similarity_chain", "transfer_optimal", "random"]`
- `spaced_repetition: bool` — Enable spaced review scheduling (default: True)
- `interleave_ratio: float` — Fraction of batch to draw from interleaved tasks (default: 0.2)
- `transfer_matrix: Optional[TransferMatrix]` — Prior knowledge of task relationships
- `max_negative_bwt: float` — Maximum tolerable backward transfer degradation (default: -0.05)

### Outputs
- `TaskSchedule` — Ordered list of (task_id, epoch_count, review_schedule) tuples
- `TransferMatrix` — Updated pairwise transfer scores after training
- `SchedulerDiagnostics` — Ordering rationale, detected anti-curriculum pairs, transfer improvements

## Acceptance Criteria

- [ ] Transfer-optimal ordering achieves >15% higher average accuracy vs. random ordering
- [ ] Spaced repetition maintains >95% retention on all tasks after 20-task sequences
- [ ] Anti-curriculum detection identifies harmful task pairs with >90% precision
- [ ] Dynamic reordering improves mid-sequence outcomes by >10% over fixed schedules
- [ ] Transfer matrix computation scales O(T²) where T is number of tasks
- [ ] Interleaving compatible with all replay strategies from Phase 33.3
- [ ] Integration tests with Phase 33.5 ContinualOrchestrator
- [ ] 95%+ test coverage, all public methods typed

## References

- Bengio et al. (2009) — Curriculum Learning
- Graves et al. (2017) — Automated Curriculum Learning
- Weinshall et al. (2018) — Curriculum Learning by Transfer Learning
- Soviany et al. (2022) — Curriculum Learning Survey
- Planning: #690

**Comment by web3guru888** [comment]
## Implementation Notes — CurriculumScheduler

### Python Skeleton

```python
from __future__ import annotations
from dataclasses import dataclass, field
from typing import Any, Dict, List, Literal, Optional, Tuple
import math
import torch
from torch import Tensor

@dataclass
class TaskSpecification:
    """Metadata for a continual learning task."""
    task_id: str
    domain: str
    difficulty: float = 0.5          # estimated difficulty [0, 1]
    embedding: Optional[Tensor] = None   # task descriptor embedding
    input_shape: Tuple[int, ...] = (784,)
    num_classes: int = 10

@dataclass
class ScheduleEntry:
    task_id: str
    epoch_count: int
    review_schedule: List[int]       # future steps at which to review
    priority: float

@dataclass
class TransferMatrix:
    """Pairwise transfer scores between tasks."""
    task_ids: List[str]
    forward_transfer: Tensor         # T×T: FWT[i,j] = benefit of i on j
    backward_transfer: Tensor        # T×T: BWT[i,j] = impact on i after j

@dataclass
class SchedulerDiagnostics:
    ordering_rationale: str
    anti_curriculum_pairs: List[Tuple[str, str]]
    estimated_avg_accuracy: float
    transfer_improvement_pct: float

class CurriculumScheduler:
    """Task sequencing for optimal continual learning."""

    def __init__(
        self,
        ordering_strategy: Literal[
            "difficulty_asc", "similarity_chain", "transfer_optimal", "random"
        ] = "transfer_optimal",
        spaced_repetition: bool = True,
        interleave_ratio: float = 0.2,
        max_negative_bwt: float = -0.05,
        review_intervals: List[int] | None = None,  # Leitner-style intervals
    ) -> None:
        self._strategy = ordering_strategy
        self._spaced = spaced_repetition
        self._interleave = interleave_ratio
        self._max_neg_bwt = max_negative_bwt
        self._intervals = review_intervals or [1, 3, 7, 15, 30]  # in task-steps
        self._transfer: Optional[TransferMatrix] = None
        self._history: List[str] = []

    # ── public API ──────────────────────────────────────────────
    def schedule(self, task_pool: List[TaskSpecification]) -> List[ScheduleEntry]:
        """Generate optimal task ordering for the pool."""
        if self._strategy == "difficulty_asc":
            ordered = sorted(task_pool, key=lambda t: t.difficulty)
        elif self._strategy == "similarity_chain":
            ordered = self._greedy_similarity_chain(task_pool)
        elif self._strategy == "transfer_optimal":
            ordered = self._transfer_optimal_ordering(task_pool)
        else:
            import random as _rng
            ordered = list(task_pool)
            _rng.shuffle(ordered)

        entries: List[ScheduleEntry] = []
        for i, task in enumerate(ordered):
            review = self._compute_review_schedule(task.task_id, i) if self._spaced else []
            entries.append(ScheduleEntry(
                task_id=task.task_id,
                epoch_count=self._estimate_epochs(task),
                review_schedule=review,
                priority=1.0 - i / len(ordered),
            ))
        return entries

    def detect_anti_curriculum(
        self, ordering: List[str], transfer: TransferMatrix
    ) -> List[Tuple[str, str]]:
        """Find consecutive task pairs with excessive negative backward transfer."""
        bad_pairs: List[Tuple[str, str]] = []
        for i in range(len(ordering) - 1):
            t_a, t_b = ordering[i], ordering[i + 1]
            idx_a = transfer.task_ids.index(t_a)
            idx_b = transfer.task_ids.index(t_b)
            bwt = transfer.backward_transfer[idx_a, idx_b].item()
            if bwt < self._max_neg_bwt:
                bad_pairs.append((t_a, t_b))
        return bad_pairs

    def update_transfer(self, task_a: str, task_b: str,
                        fwt: float, bwt: float) -> None:
        """Update transfer matrix after observing training of task_b after task_a."""
        if self._transfer is None:
            self._transfer = TransferMatrix(
                task_ids=[task_a, task_b],
                forward_transfer=torch.zeros(2, 2),
                backward_transfer=torch.zeros(2, 2),
            )
        tm = self._transfer
        if task_a not in tm.task_ids:
            self._expand_matrix(task_a)
        if task_b not in tm.task_ids:
            self._expand_matrix(task_b)
        i = tm.task_ids.index(task_a)
        j = tm.task_ids.index(task_b)
        tm.forward_transfer[i, j] = fwt
        tm.backward_transfer[i, j] = bwt

    def dynamic_reorder(self, remaining: List[TaskSpecification]) -> List[ScheduleEntry]:
        """Re-optimize remaining task order based on accumulated transfer data."""
        if self._transfer is not None:
            return self.schedule(remaining)  # uses updated transfer matrix
        return self.schedule(remaining)

    def interleave_batch(
        self, current_task_data: Tensor, previous_tasks: Dict[str, Tensor]
    ) -> Tensor:
        """Create interleaved batch from current + previous task data."""
        n_interleave = int(current_task_data.size(0) * self._interleave)
        n_current = current_task_data.size(0) - n_interleave
        if not previous_tasks:
            return current_task_data
        prev_samples = torch.cat(list(previous_tasks.values()))
        indices = torch.randperm(prev_samples.size(0))[:n_interleave]
        interleaved = prev_samples[indices]
        return torch.cat([current_task_data[:n_current], interleaved])

    # ── internals ───────────────────────────────────────────────
    def _compute_review_schedule(self, task_id: str, position: int) -> List[int]:
        return [position + iv for iv in self._intervals]

    def _estimate_epochs(self, task: TaskSpecification) -> int:
        return max(1, int(5 + 10 * task.difficulty))

    def _greedy_similarity_chain(self, pool: List[TaskSpecification]) -> List[TaskSpecification]:
        """Order tasks by chaining most-similar consecutive pairs."""
        if not pool or pool[0].embedding is None:
            return pool
        remaining = list(pool)
        chain = [remaining.pop(0)]
        while remaining:
            last_emb = chain[-1].embedding
            sims = [torch.cosine_similarity(last_emb.unsqueeze(0), t.embedding.unsqueeze(0)).item()
                    for t in remaining]
            best_idx = max(range(len(sims)), key=lambda i: sims[i])
            chain.append(remaining.pop(best_idx))
        return chain

    def _transfer_optimal_ordering(self, pool: List[TaskSpecification]) -> List[TaskSpecification]:
        """Use transfer matrix to find ordering maximizing cumulative FWT."""
        if self._transfer is None:
            return self._greedy_similarity_chain(pool)
        # greedy: pick task with highest expected FWT given previously scheduled tasks
        remaining = list(pool)
        ordered: List[TaskSpecification] = []
        while remaining:
            if not ordered:
                # start with easiest task
                easiest = min(remaining, key=lambda t: t.difficulty)
                ordered.append(easiest)
                remaining.remove(easiest)
                continue
            best, best_score = remaining[0], -float("inf")
            for task in remaining:
                if task.task_id in self._transfer.task_ids:
                    j = self._transfer.task_ids.index(task.task_id)
                    score = sum(
                        self._transfer.forward_transfer[
                            self._transfer.task_ids.index(o.task_id), j
                        ].item()
                        for o in ordered if o.task_id in self._transfer.task_ids
                    )
                    if score > best_score:
                        best, best_score = task, score
            ordered.append(best)
            remaining.remove(best)
        return ordered

    def _expand_matrix(self, task_id: str) -> None: ...
```

### Algorithm — Spaced Repetition (Leitner Adaptation)

```
procedure SPACED_REVIEW_SCHEDULE(task_id, position, intervals):
    schedule ← []
    for interval in intervals:          // e.g., [1, 3, 7, 15, 30] task-steps
        review_step ← position + interval
        schedule.append(review_step)
    return schedule

procedure SHOULD_REVIEW(current_step, all_schedules):
    tasks_to_review ← []
    for task_id, schedule in all_schedules:
        if current_step in schedule:
            tasks_to_review.append(task_id)
    return tasks_to_review
```

### Type Table (mypy-compatible)

| Symbol | Type | Description |
|--------|------|-------------|
| `ordering_strategy` | `Literal[...]` | Task ordering heuristic |
| `spaced_repetition` | `bool` | Enable Leitner-style review |
| `interleave_ratio` | `float` | Fraction of batch to interleave |
| `TransferMatrix.forward_transfer` | `Tensor` | T×T FWT matrix |
| `TransferMatrix.backward_transfer` | `Tensor` | T×T BWT matrix |
| `ScheduleEntry.review_schedule` | `List[int]` | Future review steps |
| `SchedulerDiagnostics.anti_curriculum_pairs` | `List[Tuple[str, str]]` | Harmful orderings |
| `max_negative_bwt` | `float` | BWT threshold for anti-curriculum |
